### PR TITLE
PEN-1796: Lazy Load Blocks (based on RC)

### PR DESCRIPTION
**Pull request by vinnie-olsen**
**If you have not filled out the checklist below, the pr is not ready for review.**

## Description
Implemented <LazyLoad/> SDK component in several blocks, modified/fixed unit tests

## Jira Ticket
- [PEN-1796](https://arcpublishing.atlassian.net/browse/PEN-1796)

## Acceptance Criteria
image

## Test Steps
1. In `Fusion-News-Theme` repo, checkout `master` & `git pull`
2. After checking out this feature branch in `fusion-news-theme-blocks`, run `npx lerna clean -y && rm -rf node_modules && npm i && npx lerna clean -y`
3. In your local `engine-theme-sdk` repo, checkout this branch https://github.com/WPMedia/fusion-news-theme-blocks/pull/755 and then run `rm -rf ./node_modules && npm i && npm run build`
4. In your `Fusion-News-Theme``.env`, make sure your `ENGINE_SDK_REPO` variable is set to the location of your engine SDK repo
5. Run `npx fusion start -f -l @wpmedia/ads-block,@wpmedia/header-nav-chain-block,@wpmedia/article-body-block,@wpmedia/author-bio-block,@wpmedia/footer-block,@wpmedia/gallery-block,@wpmedia/share-bar-block,@wpmedia/small-manual-promo-block,@wpmedia/medium-manual-promo-block,@wpmedia/large-manual-promo-block,@wpmedia/extra-large-manual-promo-block,@wpmedia/small-promo-block,@wpmedia/medium-promo-block,@wpmedia/large-promo-block,@wpmedia/extra-large-promo-block,@wpmedia/tag-title-block,@wpmedia/top-table-list-block,@wpmedia/header-block,@wpmedia/video-player-block,@wpmedia/tag-title-block,@wpmedia/simple-list-block,@wpmedia/numbered-list-block,@wpmedia/results-list-block,@wpmedia/results-list-block,@wpmedia/search-results-list-block,@wpmedia/card-list-block,@wpmedia/article-tag-block` in `Fusion-News-Theme`
6. Open the `all-blocks` page in PageBuilder editor
7. In your PageBuilder editor, set the `Lazy Load block?` checkbox to "checked" for all of the blocks listed in the screenshot below
8. Stage & publish and open http://localhost/all-blocks/?_website=the-gazette in a separate tab
9. You should now see that all of the affected blocks are correctly lazy-loaded. This is a bit difficult to see since it loads quickly as you scroll down the viewport. The best way I've found is to set (in Chrome DevTools) breakpoints and/or log points inside the `Render` component (`CardListRender`, `FooterRender`) that each of the affected blocks now has nested in their component bodies (to wrap the `<LazyLoad/>` component around since it has to be wrapped around a component and not just a block of JSX). You can also _sort of_ tell that it's working if you scroll down the page really fast or use the PgUp/PgDown keys. You can also monitor changes in the DOM through the `Elements` tab but I found this harder to track.

image

### Before
Lazy-loading option was only available on Ads Block previously

### After
Lazy-loading option is now available on all blocks listed in the above screenshot

## Dependencies or Side Effects
_Examples of dependencies or side effects are:_
- Additional settings that will be required in the blocks.json
- Changes to the custom fields which will require users to reconfigure features
- Update to css framework or SDK
- Dependency on another PR that needs to be merged first

## Author Checklist
_The author of the PR should fill out the following sections to ensure this PR is ready for review._
- [X] Confirmed all the test steps a reviewer will follow above are working. 
- [X] Confirmed there are no linter errors. Please run `npm run lint` to check for errors. Often, `npm run lint:fix` will fix those errors and warnings.
- [X] Ran this code locally and checked that there are not any unintended side effects. For example, that a CSS selector is scoped only to a particular block.
- [X] Confirmed this PR has reasonable code coverage. You can run `npm run test:coverage` to see your progress.
  - [X] Confirmed this PR has unit test files
  - [X] Ran `npm run test`, made sure all tests are passing
  - [ ] If the amount of work to write unit tests for this change are excessive,
please explain why (so that we can fix it whenever it gets refactored).
- [X] Confirmed relevant documentation has been updated/added.

## Reviewer Checklist 
_The reviewer of the PR should copy-paste this template into the review comments on review._

- [ ] Linting code actions have passed.
- [ ] Ran the code locally based on the test instructions.
  - [ ] I don’t think this is needed to be tested locally. For example, a padding style change (storybook?) or a logic change (write a test).
- [ ] I am a member of the engine theme team so that I can approve and merge this. If you're not on the team, you won't have access to approve and merge this pr. 
- [ ] Looked to see that the new or changed code has code coverage, specifically. We want the global code coverage to keep on going up with targeted testing.

### CODE REVIEW

**Comment by JackHowa** [comment]
@vinnie-olsen do not merge?

File: blocks/ads-block/features/ads/default.jsx
Comment (JackHowa) [inline blocks/ads-block/features/ads/default.jsx:100]: good catch

**Review by JackHowa — DISMISSED** [review]
Good work! Works as expected. If it's possible, can you add some sort of description to lazy load? Thinking as a non-technical user, I would appreciate this. A mass-edit of the proptypes could probably help

File: blocks/article-body-block/chains/article-body/default.test.jsx
Comment (JackHowa) [inline blocks/article-body-block/chains/article-body/default.test.jsx:1474]: was this removed deliberately?

Comment (vinnie-olsen) [reply blocks/article-body-block/chains/article-body/default.test.jsx:1474]: Ahhh good catch 👍  I don't remember why this was removed so I think it was accidental while I was going through and fixing all of the unit tests 🤔  I'll restore this and re-verify locally that the tests are passing

**Comment by vinnie-olsen** [comment]
Thank you @JackHowa ! Ahh the `do-not-merge` was just on here from before while I was making changes to the PR and didn't want anyone to review yet. I just forgot to remove it 🤦 

Yep I will definitely add a description for the custom field to all blocks 👍

File: blocks/article-body-block/chains/article-body/default.test.jsx
Comment (vinnie-olsen) [reply blocks/article-body-block/chains/article-body/default.test.jsx:1474]: Ohhh wait now that I'm re-testing this - I remember why I removed it. It's because this unit test is basically testing implementation details of `ImageMetadata`. It was only working previously because `ImageMetadata` wasn't yet mocked but is now so the test isn't possible to execute the same way. I only realized it when I was changing things to fix all of the tests but it seems like the Article Body block probably shouldn't be testing the exact format that `ImageMetadata` renders with right? Seems like that would be the domain of `ImageMetadata` unit tests to test that logic.

**Comment by vinnie-olsen** [comment]
@JackHowa added descriptions for the `lazyLoad` custom fields to all blocks 👍  I saw your comment on the ticket about the wording though so let me know if this looks good to merge or if we would like to wait on the team to weigh in on how the description should be written.

Also I left an explanation of why I removed the `should hide photographer and credit` unit tests in the Article Block but let me know if you disagree with that or need any further clarification 🙂  (also looks like I'll need a re-approval since I pushed a new commit 😬 )

**Comment by vinnie-olsen** [comment]
@JackHowa updated `lazyLoad` custom field description with your suggestion 👍

**Comment by vinnie-olsen** [comment]
@JackHowa Looks like I need a second approval since I pushed those changes 😬  Whenever you get a chance, could you re-approve if my recent changes look good?

**Review by JackHowa — APPROVED** [review]
Looks to be working as described. I might expect the block to be imported or loaded only when shown, like react.lazy and suspense. But I think this is the intermediary goal to reduce dom size. Thanks for adding the description so quick @vinnie-olsen

**Comment by vinnie-olsen** [comment]
No problem and thank you for the thorough review @JackHowa! 🙂 🎉

**Pull request merged**
